hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMPathsPurgeResponseWithFSO.java

```diff
@@ -0,0 +1,130 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.om.response.key;
+
+import org.apache.hadoop.hdds.utils.db.BatchOperation;
+import org.apache.hadoop.ozone.OmUtils;
+import org.apache.hadoop.ozone.om.OMMetadataManager;
+import org.apache.hadoop.ozone.om.helpers.BucketLayout;
+import org.apache.hadoop.ozone.om.helpers.OmKeyInfo;
+import org.apache.hadoop.ozone.om.helpers.RepeatedOmKeyInfo;
+import org.apache.hadoop.ozone.om.request.key.OMPathsPurgeRequestWithFSO;
+import org.apache.hadoop.ozone.om.response.CleanupTableInfo;
+import org.apache.hadoop.ozone.om.response.OMClientResponse;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos.OMResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.List;
+
+import static org.apache.hadoop.ozone.om.OmMetadataManagerImpl.DELETED_DIR_TABLE;
+import static org.apache.hadoop.ozone.om.OmMetadataManagerImpl.DELETED_TABLE;
+import static org.apache.hadoop.ozone.om.OmMetadataManagerImpl.DIRECTORY_TABLE;
+import static org.apache.hadoop.ozone.om.OmMetadataManagerImpl.FILE_TABLE;
+
+/**
+ * Response for {@link OMPathsPurgeRequestWithFSO} request.
+ */
+@CleanupTableInfo(cleanupTables = {DELETED_TABLE, DELETED_DIR_TABLE,
+    DIRECTORY_TABLE, FILE_TABLE})
+public class OMPathsPurgeResponseWithFSO extends OMClientResponse {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OMPathsPurgeResponseWithFSO.class);
+
+  private List<OzoneManagerProtocolProtos.KeyInfo> markDeletedDirList;
+  private List<String> dirList;
+  private List<OzoneManagerProtocolProtos.KeyInfo> fileList;
+  private boolean isRatisEnabled;
+
+
+  public OMPathsPurgeResponseWithFSO(@Nonnull OMResponse omResponse,
+      @Nonnull List<OzoneManagerProtocolProtos.KeyInfo> markDeletedDirs,
+      @Nonnull List<OzoneManagerProtocolProtos.KeyInfo> files,
+      @Nonnull List<String> dirs, boolean isRatisEnabled) {
+    super(omResponse);
+    this.markDeletedDirList = markDeletedDirs;
+    this.dirList = dirs;
+    this.fileList = files;
+    this.isRatisEnabled = isRatisEnabled;
+  }
+
+  @Override
+  public void addToDBBatch(OMMetadataManager omMetadataManager,
+      BatchOperation batchOperation) throws IOException {
+
+    // Add all sub-directories to deleted directory table.
+    for (OzoneManagerProtocolProtos.KeyInfo key : markDeletedDirList) {
+      OmKeyInfo keyInfo = OmKeyInfo.getFromProtobuf(key);
+      String ozoneDbKey = omMetadataManager.getOzonePathKey(
+          keyInfo.getParentObjectID(), keyInfo.getFileName());
+      omMetadataManager.getDeletedDirTable().putWithBatch(batchOperation,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code might contain a resource leak. A resource leak can slow down or crash your system.

**Fix**
Consider closing the resource returned by the *getDeletedDirTable* method in a try-finally block. Currently, there are execution paths that do not contain closure statements, for example, when Table.putWithBatch throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).